### ISSUE TITLE
Assertion caused by scan, vmap, and custom derivatives

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#4521]
It seems like there's a bad interaction between vmap and something in custom derivatives:
```
  File "/home/neil/.pyenv/versions/3.8.5/lib/python3.8/site-packages/jax/custom_derivatives.py", line 464, in __call__
    out_flat = custom_vjp_call_jaxpr(flat_fun, flat_fwd, flat_bwd,
  File "/home/neil/.pyenv/versions/3.8.5/lib/python3.8/site-packages/jax/custom_derivatives.py", line 530, in custom_vjp_call_jaxpr
    return custom_vjp_call_jaxpr_p.bind(*args, fun_jaxpr=fun_jaxpr,
  File "/home/neil/.pyenv/versions/3.8.5/lib/python3.8/site-packages/jax/core.py", line 266, in bind
    out = top_trace.process_primitive(self, tracers, params)
  File "/home/neil/.pyenv/versions/3.8.5/lib/python3.8/site-packages/jax/interpreters/batching.py", line 159, in process_primitive
    return map(partial(BatchTracer, self), val_out, dim_out)
  File "/home/neil/.pyenv/versions/3.8.5/lib/python3.8/site-packages/jax/util.py", line 34, in safe_map
    assert len(arg) == n, 'length mismatch: {}'.format(list(map(len, args)))
AssertionError: length mismatch: [13, 100]
```
I've spent about a day trying to make a minimum working example.  I'll keep trying, but just sharing this in case this jumps out as obvious to someone.

**Comment by shoyer** [comment]
Any chance you're defining your `custom_vjp` function inside another function or a class? My understanding is that `custom_vjp`/`custom_jvp` should always be written as "top level" functions (i.e., with no indentation).

Otherwise, this is definitely a bug -- the newer custom derivatives interface was designed specifically around handling `vmap` of custom derivatives properly.

ping @mattjj

**Comment by NeilGirdhar** [comment]
@shoyer I am defining a custom_vjp, but it's on a top-level function.

**Comment by NeilGirdhar** [comment]
```python
from functools import partial

import jax
from jax import grad
from jax import numpy as jnp
from jax import vjp, vmap
from jax.lax import scan

code_shape = (0,)
v_shape = (2,)

def infer(weights, rng):
    inference_rngs = jnp.ones((10, 5)) * rng
    infer_with_loss = partial(_v_infer_with_loss, inference_rngs)
    f = grad(infer_with_loss)
    weights_bar = f(weights)
    assert False

def _v_infer_with_loss(inference_rngs, weights):
    f = vmap(infer_encoding_configuration, in_axes=(0, None), out_axes=())
    primals = f(inference_rngs, weights)

def _infer_encoding_configuration(weights):
    return weights @ jnp.zeros(code_shape)

@jax.custom_vjp
def infer_encoding_configuration(initial_rng, weights):
    return _infer_encoding_configuration(weights)

def _infer_encoding_configuration_fwd(initial_rng, weights):
    v, residuals = vjp(_infer_encoding_configuration, weights)
    return v, residuals  # changing residuals to None here lets things pass 

def _infer_encoding_configuration_bwd(residuals, tpec_bar):
    assert False

infer_encoding_configuration.defvjp(_infer_encoding_configuration_fwd,
                                    _infer_encoding_configuration_bwd)

def z(rng, _):
    weights = jnp.zeros(v_shape + code_shape)
    infer(weights, rng)
    return rng, _

scan(z, jnp.ones(5), None, 100)
```

This appears to be related to https://github.com/google/jax/pull/3705

**Comment by mattjj** [comment]
Thanks, @NeilGirdhar , not only for this report and clear minimal reproduction, but also for your extreme patience putting up with `custom_vjp` bugs as I delayed actually fixing things once and for all (my excuse was landing #3370).

But we will soon awake from that long nightmare! I have a branch, just pushed to #4008 on Friday, that will close all open issues about `custom_jvp` and `custom_vjp` closure difficulties, and I think will eliminate the possibility of these bugs in the future.

When I tried this repro on #4008 I got this error instead:

> ValueError: vmap out_axes specification must be a tree prefix of the corresponding value, got specification () for value tree PyTreeDef(None, []).

Is that an error we expect?

**Comment by mattjj** [comment]
@shoyer by the way, #4008 removes the top-level-ness requirement for custom_jvp/vjp. See the tests!

**Comment by NeilGirdhar** [comment]
@mattjj No problem at all.  It's my pleasure to produce MWEs even though it's time consuming especially since the JAX team has been exceptionally fast at responding to issues, so thank you very much for that.

It looks like you've fixed the bug!  I've updated my MWE in case you want to add it or something like it to your testing.  I guess you can add "Fixes #4521" to the description of #4008 so that this will be automatically closed?

Thanks again for addressing this so fast!

**Comment by mattjj** [comment]
@NeilGirdhar Yeah perfect, I'll add your MWE as a test, and I'll add this issue to the "fixes" list.

**Comment by NeilGirdhar** [comment]
Closed by #4008
